keps/sig-network/2610-allports-services/README.md

```diff
@@ -0,0 +1,569 @@
+
+4) Endpoints and EndpointSlices controller should create Endpoints with empty port values.
+
+5) DNS Providers should handle empty ports in Services and Endpoints.
+   There will be no DNS SRV Records for Services with `allPorts` set.
+   coreDNS handles empty ports in [services](https://github.com/coredns/coredns/blob/09b63df9c1584bb5389d1b681698631bcd7c19e1/plugin/kubernetes/kubernetes.go#L577) and [endpoints.](https://github.com/coredns/coredns/blob/09b63df9c1584bb5389d1b681698631bcd7c19e1/plugin/kubernetes/kubernetes.go#L559)
+   kube-dns also handles empty ports in [services](https://github.com/kubernetes/dns/blob/077a43e83e648ba5f04bae18ffcb824edc9db967/pkg/dns/dns.go#L506) and [endpoints.](https://github.com/kubernetes/dns/blob/077a43e83e648ba5f04bae18ffcb824edc9db967/pkg/dns/dns.go#L541)
+   There is a warning in kube-dns for [empty port services](https://github.com/kubernetes/dns/blob/077a43e83e648ba5f04bae18ffcb824edc9db967/pkg/dns/dns.go#L320) that could be removed after GA of AllPorts.
+
+6) There will be no environment variable of the form "_SERVICE_PORT" for services with `allPorts` set. This [codepath](https://github.com/kubernetes/kubernetes/blob/e1f971d5c2a1002c4e90471d064f87f297740aba/pkg/kubelet/envvars/envvars.go#L48) currently assumes non-zero ports and that will be updated.
+
+#### Open Questions to be resolved:
+
+1) How should IPVS implementation be handled?
+Options are to a) use iptables for DNAT b) use ipvs rules that results in same 5-tuple requests being sent to the same backend pod. c) create allPorts services as "fwmark-service" and assign a unique mark for each of them.
+
+2) Identfy how Service Mesh(Istio)/Calico/MetalLB can support AllPorts.
+
+### Test Plan
+
+Unit tests:
+
+* To verify API validation of the `allPorts` and `ports` fields.
+* To verify that all users(kube-proxy, kubelet, various controllers) of Service/Endpoints can handle empty ports.
+* To check the default value of `allPorts` on each type of Service.
+
+E2E tests:
+
+* To verify that default behavior of `allPorts` does not break any existing e2e tests.
+* Test setting `allPorts` on a new Service and connecting to the service VIP on a few different ports.
+* Test setting `allPorts` on an existing Service and connecting to the service VIP on a few different ports.
+* Test unsetting `allPorts` on a service and specifying a single port allows traffic only for that port.
+* Test setting `allPorts` explicitly to false.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Add a new field `allPorts` to Service, but it can only be set when the feature gate is on.
+
+
+#### Alpha -> Beta Graduation
+
+- Ensure that the main clients of Service/Endpoints API - kube-proxy, kubelet, loadbalancer controllers have added support to handle empty ports. All supported node versions for the given master version(that graduates `allPorts` to Beta) should handle this case correctly.
+- Tests are in Testgrid and linked in KEP
+
```

Comment:
do we want to gate on implementation of the feature on external components and cloud providers? 
_at least X major cloud provider implements the feature_